book-content/chapters/06-objects.md

```diff
@@ -0,0 +1,1386 @@
+
+We can choose any name for the `index`. It's just a description.
+
+```typescript
+const albumAwards: {
+  [iCanBeAnything: string]: boolean;
+} = {};
+```
+
+The same syntax can also be used with types and interfaces:
+
+```typescript
+interface AlbumAwards {
+  [index: string]: boolean;
+}
+
+const beyonceAwards: AlbumAwards = {
+  Grammy: true,
+  Billboard: true,
+};
+```
+
+Index signatures are one way to handle dynamic keys. But there's a utility type that some argue is even better.
+
+### Using a Record Type for Dynamic Keys
+
+The `Record` utility type is another option for supporting dynamic keys.
+
+Here's how we would use `Record` for the `albumAwards` object, where the key will be a string and the value will be a boolean:
+
+```typescript
+const albumAwards: Record<string, boolean> = {};
+
+albumAwards.Grammy = true;
+```
+
+The first type argument is the key, and the second type argument is the value. This is a more concise way to achieve a similar result as an index signature.
+
+`Record` can also support a union type as keys, but an index signature can't:
+
+```ts twoslash
+// @errors: 1337
+const albumAwards1: Record<"Grammy" | "MercuryPrize" | "Billboard", boolean> = {
+  Grammy: true,
+  MercuryPrize: false,
+  Billboard: true,
+};
+
+const albumAwards2: {
+  [index: "Grammy" | "MercuryPrize" | "Billboard"]: boolean;
+} = {
+  Grammy: true,
+  MercuryPrize: false,
+  Billboard: true,
+};
+```
+
+Index signatures can't use literal types, but `Record` can. We'll look at why this is when we explore mapped types in a later chapter.
+
+The `Record` type helper is a repeatable pattern that's easy to read and understand, and is a bit more flexible than an index signature. It's my go-to for dynamic keys.
+
+### Combining Known and Dynamic Keys
+
+In many cases there will be a base set of keys we know we want to include, but we also want to allow for additional keys to be added dynamically.
+
+For example, say we are working with a base set of awards we know were nominations, but we don't know what other awards are in play. We can use the `Record` type to define a base set of awards and then use an intersection to extend it with an index signature for additional awards:
+
+```typescript
+type BaseAwards = "Grammy" | "MercuryPrize" | "Billboard";
+
+type ExtendedAlbumAwards = Record<BaseAwards, boolean> & {
+  [award: string]: boolean;
+};
+
+const extendedNominations: ExtendedAlbumAwards = {
+  Grammy: true,
+  MercuryPrize: false,
+  Billboard: true, // Additional awards can be dynamically added
+  "American Music Awards": true,
+};
+```
+
+This technique would also work when using an interface and the `extends` keyword:
+
+```typescript
+interface BaseAwards {
+  Grammy: boolean;
+  MercuryPrize: boolean;
+  Billboard: boolean;
+}
+
+interface ExtendedAlbumAwards extends BaseAwards {
+  [award: string]: boolean;
+}
+```
+
+This version is preferable because, in general, `interface extends` is preferable to intersections.
```

Comment:
it's even more important with index signatures because to make sound types all of ur values have to be compatible with applicable index signatures. Intersections won't warn you about this at all.

People often ask to be able to break this rule but that just creates problems: [TS playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEYD2A7AzgF3kgRgKwC54BveAbQGsjMYBLFAcwF0ickkIQoV4BfeAGQl4FFEgDuKIigCuAWxwgYfANwAoNcnRY4aGRAwBGeAF5s+AHSiJKdQHo78J-AB6Afg1bMIsZOoY6RlN4ACJrSRD1Lx0QPQMAJmDcPEpfFCZ7R2d3NSA)